x-pack/plugins/infra/public/pages/logs/settings/form_elements.tsx

```diff
@@ -0,0 +1,243 @@
+  childFormElements,
+  validate,
+}: {
+  childFormElements: FormElementMap<FormValues>;
+  validate?: (values: FormValues) => Promise<InvalidReason[]>;
+}): CompositeFormElement<FormValues, InvalidReason> => {
+  const childFormElementEntries = useMemo(
+    () => Object.entries(childFormElements) as ObjectEntries<typeof childFormElements>,
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    Object.entries(childFormElements).flat()
+  );
+
+  const value = useMemo(
+    () =>
+      childFormElementEntries.reduce<FormValues>(
+        (accumulatedFormValues, [formElementName, formElement]) => ({
+          ...accumulatedFormValues,
+          [formElementName]: formElement.value,
+        }),
+        {} as FormValues
+      ),
+    [childFormElementEntries]
+  );
+
+  const updateValue = useCallback(
+    (updater: (previousValues: FormValues) => FormValues) => {
+      const newValues = updater(value);
+
+      childFormElementEntries.forEach(([formElementName, formElement]) =>
+        formElement.updateValue(() => newValues[formElementName])
+      );
+    },
+    [childFormElementEntries, value]
+  );
+
+  const isDirty = useMemo(
+    () => childFormElementEntries.some(([, formElement]) => formElement.isDirty),
+    [childFormElementEntries]
+  );
+
+  const formValidity = useValidity(value, validate);
+  const childFormElementsValidity = useMemo<
+    FormElementValidity<InvalidReason | ChildFormValidationError>
+  >(() => {
+    if (
+      childFormElementEntries.some(([, formElement]) => formElement.validity.validity === 'invalid')
+    ) {
+      return {
+        validity: 'invalid',
+        reasons: [{ type: 'child' }],
+      };
+    } else if (
+      childFormElementEntries.some(([, formElement]) => formElement.validity.validity === 'pending')
+    ) {
+      return {
+        validity: 'pending',
+      };
+    } else {
+      return {
+        validity: 'valid',
+      };
+    }
+  }, [childFormElementEntries]);
+
+  const validity = useMemo(() => getCombinedValidity(formValidity, childFormElementsValidity), [
+    formValidity,
+    childFormElementsValidity,
+  ]);
+
+  const resetValue = useCallback(() => {
+    childFormElementEntries.forEach(([, formElement]) => formElement.resetValue());
+  }, [childFormElementEntries]);
+
+  const initialValue = useMemo(
+    () =>
+      childFormElementEntries.reduce<FormValues>(
+        (accumulatedFormValues, [formElementName, formElement]) => ({
+          ...accumulatedFormValues,
+          [formElementName]: formElement.initialValue,
+        }),
+        {} as FormValues
+      ),
+    [childFormElementEntries]
+  );
+
+  return useMemo(
+    () => ({
+      childFormElements,
+      initialValue,
+      isDirty,
+      resetValue,
+      updateValue,
+      validity,
+      value,
+    }),
+    [childFormElements, initialValue, isDirty, resetValue, updateValue, validity, value]
+  );
+};
+
+const useValidity = <Value, InvalidReason>(
```

Comment:
:information_source: This turns the async validation state into something that can be used in components.